File: listas.ml

(** val length : 'a1 list -> int **)

let rec length = function
| [] -> 0
| _::ls' -> succ (length ls')

(** val app : 'a1 list -> 'a1 list -> 'a1 list **)

let rec app ls1 ls2 =
  match ls1 with
  | [] -> ls2
  | x::ls1' -> x::(app ls1' ls2)

(** val rev : 'a1 list -> 'a1 list **)

let rec rev l = match l with
| [] -> l
| h::t -> app (rev t) (h::[])

File: listas.hs
module Listas where

import qualified Prelude

length :: ([] a1) -> Prelude.Int
length ls =
  case ls of {
   [] -> 0;
   (:) _ ls' -> Prelude.succ (length ls')}

app :: ([] a1) -> ([] a1) -> [] a1
app ls1 ls2 =
  case ls1 of {
   [] -> ls2;
   (:) x ls1' -> (:) x (app ls1' ls2)}

rev :: ([] a1) -> [] a1
rev l =
  case l of {
   [] -> l;
   (:) h t -> app (rev t) ((:) h [])}

